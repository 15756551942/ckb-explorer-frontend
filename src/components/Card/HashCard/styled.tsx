import styled from 'styled-components'

export const HashCardPanel = styled.div`
  width: 100%;
  height: auto;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  box-shadow: 2px 2px 6px 0 #dfdfdf;
  border-radius: 6px;
  background-color: #ffffff;
  padding: 0 40px;
  @media (max-width: 750px) {
    padding: 0 16px;
    border-radius: 3px;
    box-shadow: 1px 1px 3px 0 #dfdfdf;
  }
  .hash__card__content__panel {
    width: 100%;
    height: 80px;
    display: flex;
    flex-direction: row;
    align-items: center;
    overflow: hidden;
    position: relative;
    @media (max-width: 750px) {
      height: auto;
      flex-direction: ${(props: { isColumn: boolean }) => (props.isColumn ? 'column' : 'row')};
      align-items: ${(props: { isColumn: boolean }) => (props.isColumn ? 'flex-start' : 'center')};
      padding-top: 8px;
      padding-bottom: 8px;
    }
  }
  .hash__icon {
    width: 40px;
    height: 40px;
    margin-right: 8px;
  }
  .hash__title {
    font-size: 26px;
    font-weight: 600;
    color: #000000;
    white-space: nowrap;
    @media (max-width: 750px) {
      font-size: 15px;
    }
  }
  .hash__card__hash__content {
    display: flex;
    flex-direction: row;
    align-items: center;
  }
  #hash__text {
    margin-left: 20px;
    font-size: 18px;
    color: #000000;
    transform: translateY(3px);
    @media (max-width: 750px) {
      font-size: 13px;
      margin-left: ${(props: { isColumn: boolean }) => (props.isColumn ? '0px' : '10px')};
      font-weight: 500;
      transform: translateY(1px);
    }
  }
  .hash__copy_icon {
    cursor: pointer;
    margin-left: 20px;
    transform: translateY(6px);
    @media (max-width: 750px) {
      margin-left: 10px;
      transform: translateY(3px);
    }
    > img {
      width: 21px;
      height: 24px;
      @media (max-width: 750px) {
        width: 16px;
        height: 18px;
        margin-bottom: 3px;
      }
    }
  }
  a {
    margin-left: 10px;
    font-size: 18px;
    margin-top: 8px;
    color: ${props => props.theme.primary};
    @media (max-width: 750px) {
      margin-left: 5px;
      font-size: 12px;
      margin-top: 4px;
    }
  }
  a:hover {
    color: ${props => props.theme.primary};
  }
  #hash__value {
    color: #ffffff;
    position: absolute;
    bottom: -30px;
  }
`

export const LoadingPanel = styled.div`
  width: 100%;
`
