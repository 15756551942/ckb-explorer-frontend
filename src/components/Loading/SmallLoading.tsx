import React from 'react'
import styled from 'styled-components'
import LoadingImage from '../../assets/loading.gif'

const Loading = styled.div`
  margin: 15px 0;
  text-align: center;

  > img {
    width: 135px;
    height: 39px;
  }

  @media (max-width: 700px) {
    margin: 8px 0;
    > img {
      width: 68px;
      height: 20px;
    }
  }
`

export default () => {
  return (
    <Loading>
      <img src={LoadingImage} alt="loading" />
    </Loading>
  )
}
