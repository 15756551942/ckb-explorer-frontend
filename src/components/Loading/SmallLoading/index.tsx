import React from 'react'
import LoadingImage from '../../../assets/loading.gif'
import LoadingBlueImage from '../../../assets/blue_loading.gif'
import { isMainnet } from '../../../utils/chain'
import { SmallLoadingPanel } from './styled'

export default () => {
  return (
    <SmallLoadingPanel>
      <img src={isMainnet() ? LoadingImage : LoadingBlueImage} alt="loading" />
    </SmallLoadingPanel>
  )
}
