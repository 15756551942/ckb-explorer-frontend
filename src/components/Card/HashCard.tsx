import React, { useLayoutEffect, useState, Dispatch, SetStateAction, useContext } from 'react'
import styled from 'styled-components'
import CopyIcon from '../../assets/copy.png'
import i18n from '../../utils/i18n'
import { isMobile } from '../../utils/screen'
import { startEndEllipsis, adaptPCEllipsis } from '../../utils/string'
import { copyElementValue } from '../../utils/util'
import { AppDispatch, AppActions } from '../../contexts/providers/reducer'
import { AppContext } from '../../contexts/providers'
import SmallLoading from '../Loading/SmallLoading'

const HashCardPanel = styled.div`
  width: 100%;
  border-radius: 6px;
  box-shadow: 2px 2px 6px 0 #dfdfdf;
  background-color: #ffffff;
  height: 80px;
  display: flex;
  flex-direction: row;
  align-items: center;
  overflow: hidden;
  position: relative;

  @media (max-width: 700px) {
    height: 50px;
    border-radius: 3px;
    box-shadow: 1px 1px 3px 0 #dfdfdf;
  }

  .address_hash__title {
    margin-left: 40px;
    font-size: 30px;
    font-weight: 500;
    color: #000000;
    white-space: nowrap;

    @media (max-width: 700px) {
      font-size: 15px;
      margin-left: 20px;
    }
  }

  #address_hash__text {
    margin-left: 20px;
    font-size: 20px;
    color: #000000;
    transform: translateY(3px);

    @media (max-width: 700px) {
      font-size: 13px;
      margin-left: 10px;
      font-weight: 500;
      transform: translateY(1px);
    }
  }

  .address_hash__copy_icon {
    cursor: pointer;
    margin-left: 20px;
    transform: translateY(6px);

    @media (max-width: 700px) {
      margin-left: 10px;
      transform: translateY(3px);
    }

    > img {
      width: 21px;
      height: 24px;

      @media (max-width: 700px) {
        width: 16px;
        height: 18px;
      }
    }
  }

  #address_hash__value {
    color: #ffffff;
    position: absolute;
    bottom: -30px;
  }
`

const LoadingPanel = styled.div`
  width: 100%;
`

const handleHashText = (hash: string, isMobileDeivce: boolean, setHashText: Dispatch<SetStateAction<string>>) => {
  if (!isMobileDeivce) {
    setHashText(hash)
    return
  }
  const contentElement = document.getElementById('address_hash_content')
  const hashElement = document.getElementById('address_hash__text')
  if (hashElement && contentElement) {
    const contentReact = contentElement.getBoundingClientRect()
    const hashReact = hashElement.getBoundingClientRect()
    const textWidth = contentReact.width - hashReact.left - 16 - 20
    const textLength = Math.round(textWidth / 8.0)
    const startLength = Math.round(textLength / 2)
    const text = startEndEllipsis(hash, textLength - startLength, startLength)
    setHashText(text)
  }
}

export default ({
  title,
  hash,
  dispatch,
  loading,
}: {
  title: string
  hash: string
  dispatch: AppDispatch
  loading?: boolean
}) => {
  const [hashText, setHashText] = useState(hash)
  const isMobileDeivce = isMobile()
  const { app } = useContext(AppContext)

  // render again when language and title change
  useLayoutEffect(() => {
    if (app.language && title) {
      handleHashText(hash, isMobileDeivce, setHashText)
    }
  }, [app.language, hash, title, isMobileDeivce])

  return (
    <HashCardPanel id="address_hash_content">
      <div className="address_hash__title">{title}</div>
      {loading ? (
        <LoadingPanel>
          <SmallLoading />
        </LoadingPanel>
      ) : (
        <div id="address_hash__text">
          <code>{adaptPCEllipsis(hashText, 18, 30)}</code>
        </div>
      )}
      <div
        className="address_hash__copy_icon"
        role="button"
        tabIndex={-1}
        onKeyDown={() => {}}
        onClick={() => {
          copyElementValue(document.getElementById('address_hash__value'))
          dispatch({
            type: AppActions.ShowToastMessage,
            payload: {
              message: i18n.t('common.copied'),
            },
          })
        }}
      >
        {!loading && <img src={CopyIcon} alt="copy" />}
      </div>
      <div id="address_hash__value">{hash}</div>
    </HashCardPanel>
  )
}
