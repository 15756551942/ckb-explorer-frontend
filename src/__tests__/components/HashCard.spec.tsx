import React from 'react'
import { shallow } from 'enzyme'
import HashCard from '../../components/Card/HashCard'
import { Link } from 'react-router-dom'
import renderer from 'react-test-renderer'
import SmallLoading from '../../components/Loading/SmallLoading'

describe('HashCard Component', () => {
  it('shallow renders', () => {
    const wrapper = renderer.create(<HashCard title="hash title" hash="0xa18fdb8a" />).toJSON()
    expect(wrapper).toMatchSnapshot()
  })

  it('shallow renders', () => {
    const wrapper = renderer.create(<HashCard title="hash title" hash="0xa18fdb8a" loading />).toJSON()
    expect(wrapper).toMatchSnapshot()
  })

  it('Component Render', () => {
    const wrapper = shallow(<HashCard title="hash title" hash="0xa18fdb8a" />)
    expect(wrapper).toBeDefined()
    expect(wrapper.find('.hash__title')).toHaveLength(1)
    expect(wrapper.find('.hash__title').text()).toBe('hash title')
    expect(wrapper.find('#hash__text')).toHaveLength(1)
    expect(wrapper.find('.hash__copy_icon')).toHaveLength(1)
  })

  it('Component Render with loading', () => {
    const wrapper = shallow(<HashCard title="hash title" hash="0xa18fdb8a" loading />)
    expect(wrapper.find('#hash__text')).toHaveLength(0)
    expect(wrapper.find(SmallLoading)).toHaveLength(1)
  })

  it('Component Render with specialAddress', () => {
    const wrapper = shallow(<HashCard title="hash title" hash="0xa18fdb8ac5" specialAddress="0xa18fdb8" />)

    expect(wrapper.find(Link)).toHaveLength(1)
    expect(wrapper.find(Link).props().to).toBe('/address/0xa18fdb8')
  })
})
