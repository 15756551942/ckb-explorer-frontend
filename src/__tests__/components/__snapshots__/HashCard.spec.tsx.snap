// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HashCard Component shallow renders 1`] = `
<div
  className="sc-AxirZ gmvWQb"
>
  <div
    className="hash__card__content__panel"
    id="hash_content"
  >
    <div
      className="hash__title"
    >
      hash title
    </div>
    <div
      className="hash__card__hash__content"
    >
      <div
        className="monospace"
        id="hash__text"
      >
        <span>
          0xa18fdb8a
        </span>
      </div>
      <div
        className="sc-AxhCb dnXiiO hash__copy_icon"
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseOver={[Function]}
        role="button"
        tabIndex={-1}
      >
        <img
          alt="copy"
          src="copy.png"
        />
      </div>
    </div>
    
    <div
      id="hash__value"
    >
      0xa18fdb8a
    </div>
  </div>
</div>
`;

exports[`HashCard Component shallow renders 2`] = `
<div
  className="sc-AxirZ gmvWQb"
>
  <div
    className="hash__card__content__panel"
    id="hash_content"
  >
    <div
      className="hash__title"
    >
      hash title
    </div>
    <div
      className="hash__card__hash__content"
    >
      <div
        className="sc-AxiKw hSLCUF"
      >
        <div
          className="sc-AxjAm icQYJv"
        >
          <img
            alt="loading"
            src="loading.gif"
          />
        </div>
      </div>
      <div
        className="sc-AxhCb dnXiiO hash__copy_icon"
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseOver={[Function]}
        role="button"
        tabIndex={-1}
      />
    </div>
    
    <div
      id="hash__value"
    >
      0xa18fdb8a
    </div>
  </div>
</div>
`;
